packages/contracts/contracts/protocol/incentives/IncentivesController2.sol

```diff
@@ -0,0 +1,273 @@
+                IAToken(assets[i]).getScaledUserBalanceAndSupply(user);
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        return userState;
+    }
+
+    /**
+     * @dev Returns the total of rewards of an user, already accrued + not yet accrued
+     * @param assets List of incentivized assets to check the rewards for
+     * @param user The address of the user
+     *
+     */
+    function getPendingRewards(address[] calldata assets, address user)
+        external
+        view
+        override
+        returns (address[] memory, uint256[] memory)
+    {
+        address[] memory rewards = _allRewards;
+        uint256 rewardsLength = rewards.length;
+        uint256[] memory amounts = new uint256[](rewardsLength);
+        DistributionTypes.UserAssetState[] memory balanceData = _getUserState(assets, user);
+
+        // cant cache because of the stack too deep
+        for (uint256 i; i < assets.length;) {
+            address asset = assets[i];
+
+            for (uint256 j; j < rewardsLength;) {
+                DistributionTypes.Reward storage reward = _incentivizedAssets[asset].rewardData[_allRewards[j]];
+                amounts[j] += reward.users[user].accrued
+                    + _getReward(
+                        balanceData[i].userBalance,
+                        _getAssetIndex(reward, balanceData[i].totalSupply, _incentivizedAssets[asset].decimals),
+                        reward.users[user].index,
+                        _incentivizedAssets[asset].decimals
+                    );
+                unchecked {
+                    ++j;
+                }
+            }
+            unchecked {
+                ++i;
+            }
+        }
+
+        return (rewards, amounts);
+    }
+
+    /**
+     * @dev Claims all available rewards on the given incentivized assets
+     * @param incentivizedAssets The list of incentivized asset addresses
+     * @param to The address to send the claimed funds to
+     * @return rewards The list of possible reward addresses
+     * @return amounts The list of amounts of each reward claimed
+     *
+     */
+    function claimAllRewards(address[] calldata incentivizedAssets, address to)
+        external
+        override
+        returns (address[] memory, uint256[] memory)
+    {
+        address[] memory rewards = _allRewards;
+        uint256[] memory amounts = new uint256[](rewards.length);
+        uint256[] memory penaltyAmounts = new uint256[](rewards.length);
+
+        DistributionTypes.UserAssetState[] memory userState = _getUserState(incentivizedAssets, msg.sender);
+        _batchUpdate(msg.sender, userState);
+
+        uint256 veTotalSupply = IVeVmex(VE_VMEX).totalSupply();
+        uint256 veUserBalance = IVeVmex(VE_VMEX).balanceOf(msg.sender);
+
+        uint256 aTokenTotalSupply;
+        uint256 boostedAmount;
+        for (uint256 i; i < incentivizedAssets.length;) {
+            address asset = incentivizedAssets[i];
```

Comment:
cost of reading length and incentivizedAsset array member is roughly the same. we read length only once per iteration, while we read asset 3 times, so doing this would increase gas consumption